c++/src/Statistics.cc

```diff
@@ -279,20 +277,55 @@ namespace orc {
   }
 
   TimestampColumnStatisticsImpl::TimestampColumnStatisticsImpl
-  (const proto::ColumnStatistics& pb, bool correctStats) {
+  (const proto::ColumnStatistics& pb, const StatContext& statContext) {
     valueCount = pb.numberofvalues();
-    if (!pb.has_timestampstatistics() || !correctStats) {
+    if (!pb.has_timestampstatistics() || !statContext.correctStats) {
       _hasMinimum = false;
       _hasMaximum = false;
+      _hasLowerBound = false;
+      _hasUpperBound = false;
       minimum = 0;
       maximum = 0;
+      lowerBound = 0;
+      upperBound = 0;
     }else{
       const proto::TimestampStatistics& stats = pb.timestampstatistics();
-      _hasMinimum = stats.has_minimum();
-      _hasMaximum = stats.has_maximum();
-
-      minimum = stats.minimum();
-      maximum = stats.maximum();
+      _hasMinimum = stats.has_minimumutc() || (stats.has_minimum() && (statContext.writerTimezone != NULL));
+      _hasMaximum = stats.has_maximumutc() || (stats.has_maximum() && (statContext.writerTimezone != NULL));
+      _hasLowerBound = stats.has_minimumutc() || stats.has_minimum();
+      _hasUpperBound = stats.has_maximumutc() || stats.has_maximum();
+
+      // Timestamp stats are stored in milliseconds
+      if (stats.has_minimumutc()) {
+        minimum = stats.minimumutc();
+        lowerBound = minimum;
+      } else if (statContext.writerTimezone) {
+        int64_t writerTimeSec = stats.minimum() / 1000;
+        // multiply the offset by 1000 to convert to millisecond
+        minimum = stats.minimum() + (statContext.writerTimezone->getVariant(writerTimeSec).gmtOffset) * 1000;
+        lowerBound = minimum;
+      } else {
+        minimum = 0;
+        // subtract 1 day 1 hour (25 hours) in milliseconds to handle unknown TZ and daylight savings
```

Comment:
@omalley Note the change here to 25 hours. The additional 1 hour is to account for daylight savings.